apps/hash-ai-worker-ts/src/activities/flow-activities/research-entities-action/link-follower-agent.ts

```diff
@@ -0,0 +1,406 @@
+  const { url } = params;
+  const urlHeadFetch = await fetch(url, { method: "HEAD" });
+
+  /**
+   * Only check the content type of the URL if the HEAD request was successful.
+   *
+   * This may be because the web page requires an authenticated user to access it.
+   */
+  if (urlHeadFetch.ok) {
+    const contentType = urlHeadFetch.headers.get("Content-Type");
+
+    if (contentType && contentType.includes("application/pdf")) {
+      return true;
+    }
+  }
+  return false;
+};
+
+const exploreResource = async (params: {
+  input: LinkFollowerAgentInput;
+  resource: ResourceToExplore;
+}): Promise<
+  | {
+      status: "ok";
+      resource: ResourceToExplore;
+      possibleNextLinks: Link[];
+      inferredFacts: Fact[];
+      inferredFactsAboutEntities: LocalEntitySummary[];
+    }
+  | {
+      status: "not-explored";
+      resource: ResourceToExplore;
+      reason: string;
+    }
+> => {
+  const { resource, input } = params;
+
+  logger.debug(`Exploring resource at URL: ${resource.url}`);
+
+  const { stepId } = await getFlowContext();
+
+  const isResourcePdfFile = await isContentAtUrlPdfFile({
+    url: resource.url,
+  });
+
+  let content = "";
+
+  if (isResourcePdfFile) {
+    const { vectorStoreIndex } = await indexPdfFile({
+      fileUrl: resource.url,
+    });
+
+    const queryEngine = vectorStoreIndex.asQueryEngine({
+      retriever: vectorStoreIndex.asRetriever({
+        // Get the 15 most similar nodes
+        similarityTopK: 15,
+      }),
+    });
+
+    logProgress([
+      {
+        recordedAt: new Date().toISOString(),
+        stepId,
+        type: "ViewedFile",
+        fileUrl: resource.url,
+        explanation: resource.reason,
+      },
+    ]);
+
+    const query = dedent(`
+      Description: ${resource.descriptionOfExpectedContent}
+      Example text: ${resource.exampleOfExpectedContent}
+    `);
+
+    logger.debug(`Querying PDF file with query: ${query}`);
+
+    const { sourceNodes } = await queryEngine.query({
+      query,
+    });
+
+    const textChunks = sourceNodes?.map(({ node }) =>
+      node.getContent(MetadataMode.NONE),
+    );
+
+    if (!textChunks || textChunks.length === 0) {
+      return {
+        status: "not-explored",
+        resource,
+        reason: "No relevant sections found in the PDF.",
+      };
+    }
+
+    logger.debug(
+      `Vector DB query returned ${textChunks.length} chunks: ${stringify(
+        textChunks,
+      )}`,
+    );
+
+    const filteredAndRankedTextChunksResponse =
+      await filterAndRankTextChunksAgent({
+        description: resource.descriptionOfExpectedContent,
+        exampleText: resource.exampleOfExpectedContent,
+        textChunks,
+      });
+
+    if (
+      filteredAndRankedTextChunksResponse.status !== "ok" ||
+      filteredAndRankedTextChunksResponse.orderedRelevantTextChunks.length === 0
+    ) {
+      /** @todo: consider improving the error reporting of this */
```

Comment:
can you update the comment to specify in what way(s) we might do this?

Reply:
What I'm referring to here is that we're handling errors encountered in the `filterAndRankTextChunksAgent` method and a valid response with no relevant chunks gracefully in the same way, we may want to handle errors differently here. I've clarified this a bit in the todo in https://github.com/hashintel/hash/pull/4649/commits/614571eed27ad792b04ab5734d5d5089b5ba94f6, and added an error log in https://github.com/hashintel/hash/pull/4649/commits/868fac173852dad2928d14777a35f29b7cbb5a83 (which won't impact how this is handled at runtime, as we will still proceed as if there were no relevant text chunks).